### PR TITLE
Restructure Labels topic

### LINKED ISSUE #17770: Apply custom metadata
This is just hanging there in an odd place.

http://docs.docker.com/engine/userguide/labels-custom-metadata/

### PR DESCRIPTION
**Pull request by mdlinville**
Fixes #17770 

**\- What I did**

A few points of work:
- Took the topic out of the left-hand nav and made it reachable from the User guide intro (addresses #17770 directly)
- Condensed the topic's contents, presenting only conceptual info and pointing
  instead to the command-line references for each type of object
- Added some info and a link to the new topics from the Dockerfile reference since it did not mention labels at all

A big part of the point is to establish a pattern of thinking and use around how
Docker uses labels and what they mean in different contexts.

Signed-off-by: Misty Stanley-Jones misty@docker.com

PTAL @thaJeztah @LK4D4 
CC @dperny @tonistiigi

### CODE REVIEW

File: docs/userguide/eng-image/dockerfile_best-practices.md [hunk 1]
```diff
@@ -142,7 +173,7 @@ As always, to make your `Dockerfile` more readable, understandable, and
 maintainable, split long or complex `RUN` statements on multiple lines separated
 with backslashes.
 
-### apt-get
+#### apt-get
```
Comment (mstanleyjones) [inline docs/userguide/eng-image/dockerfile_best-practices.md]: This was incidental, but it seemed like if RUN was a third-level heading, and apt-get was a RUN example, it should be a 4th-level heading. That's why the change is here even though it is not really relevant to this PR.

File: docs/userguide/eng-image/dockerfile_best-practices.md [hunk 2]
```diff
@@ -134,6 +134,37 @@ image. We recommend the [Debian image](https://hub.docker.com/_/debian/)
...
+
+You can add labels to your image to help organize images by project, record
+licensing information, to aid in automation, or for other reasons. For each
+label, add a line beginning with `LABEL` and with one or more key-value pairs.
+The following examples show the different acceptable formats.
+
+```nohighlight
+# Set one or more individual labels
+LABEL com.example.version="0.0.1-beta"
+LABEL vendor=ACME\ Incorporated
```
Comment (tonistiigi) [inline docs/userguide/eng-image/dockerfile_best-practices.md]: It's a bit weird that we use quotes where they are not strictly needed but not when the value contains a space. I know we are trying to show different ways here but I think we should only promote quoted versions if spaces are needed in value as they are much more readable.

File: docs/userguide/labels-custom-metadata.md [hunk 3]
```diff
@@ -1,230 +1,106 @@
 <!--[metadata]>
 +++
-title = "Apply custom metadata"
-description = "Learn how to work with custom metadata in Docker, using labels."
-keywords = ["Usage, user guide, labels, metadata, docker, documentation, examples,  annotating"]
+title = "Managing Docker Object Labels"
```
Comment (tonistiigi) [inline docs/userguide/labels-custom-metadata.md]: Why the capitalized format? Don't see it in other files.

File: docs/userguide/intro.md [hunk 4]
```diff
@@ -64,6 +65,24 @@ learning how to manage data, volumes and mounts inside our containers.
 
 Go to [Managing Data in Containers](../tutorials/dockervolumes.md).
 
+## Managing Metadata (Labels) for Docker Objects
+
+_Labels_ are a mechanism for applying metadata to Docker objects, including:
+
+- Images
+- Containers
+- Daemons
```
Comment (tonistiigi) [inline docs/userguide/intro.md]: I'd use `Local daemon`. There is no concept of multiple daemons, then they are already swarm nodes.

**Comment by AkihiroSuda** [comment]
I think it'd be better if there is a description about that only labels for `node`s and `service`s can be dynamically updated, but others not

File: docs/userguide/labels-custom-metadata.md [hunk 5]
```diff
@@ -1,230 +1,106 @@
...
+  - [Adding labels to volumes](../reference/commandline/volume_create.md)
+  - [Inspecting a volume's labels](../reference/commandline/volume_inspect.md)
+  - [Filtering volumes by label](../reference/commandline/volume_ls.md#filtering)
+
+- Networks
+  - [Adding labels to a network](../reference/commandline/network_create.md)
+  - [Inspecting a network's labels](../reference/commandline/network_inspect.md)
+  - [Filtering networks by label](../reference/commandline/network_ls.md#filtering)
+
+- Swarm Nodes
```
Comment (AkihiroSuda) [inline docs/userguide/labels-custom-metadata.md]: and Services?

File: docs/userguide/labels-custom-metadata.md (hunk 3 above)
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: Yes, we standardize on sentence capitalization; see the style guide; https://docs.docker.com/opensource/doc-style/#/in-headings

File: docs/userguide/eng-image/dockerfile_best-practices.md (hunk 2 above)
Comment (mstanleyjones) [inline docs/userguide/eng-image/dockerfile_best-practices.md]: I agree. I pasted that from already-existing content and it was already that way, and maybe they were trying to show that you need to either quote or escape. But it seems to me that always quoting is safest. Then the only oddity is if your string contains quotes, in which case I suppose you would escape the inner quote.

File: docs/userguide/labels-custom-metadata.md (hunk 3 above)
Comment (mstanleyjones) [inline docs/userguide/labels-custom-metadata.md]: Thanks, I'll change it.

File: docs/userguide/intro.md (hunk 4 above)
Comment (mstanleyjones) [inline docs/userguide/intro.md]: Noted.

File: docs/userguide/labels-custom-metadata.md (hunk 5 above)
Comment (mstanleyjones) [inline docs/userguide/labels-custom-metadata.md]: Thanks!

**Comment by mstanleyjones** [comment]
All feedback so far addressed.

File: docs/userguide/eng-image/dockerfile_best-practices.md [hunk 6]
```diff
@@ -134,6 +134,41 @@ image. We recommend the [Debian image](https://hub.docker.com/_/debian/)
 since it’s very tightly controlled and kept minimal (currently under 150 mb),
 while still being a full distribution.
 
+### LABEL
+
+(Understanding Docker Object Labels)(../labels-custom-metadata.md)
```
Comment (thaJeztah) [inline docs/userguide/eng-image/dockerfile_best-practices.md]: s/Understanding Docker Object Labels/Understanding Docker object labels/

File: docs/userguide/eng-image/dockerfile_best-practices.md [hunk 7]
```diff
@@ -134,6 +134,41 @@ image. We recommend the [Debian image](https://hub.docker.com/_/debian/)
...
+You can add labels to your image to help organize images by project, record
+licensing information, to aid in automation, or for other reasons. For each
+label, add a line beginning with `LABEL` and with one or more key-value pairs.
+The following examples show the different acceptable formats.
+
+>**Note**: If your string contains spaces, it must be quoted **or** the spaces
+must be escaped. If your string contains inner quote characters (`"`), escape
+them as well.
+
+```nohighlight
```
Comment (thaJeztah) [inline docs/userguide/eng-image/dockerfile_best-practices.md]: Does this work currently (`nohighlight`?). I know we talked about implementing, so wondering if that was done

In this case, given that this is a Dockerfile example, perhaps better to change to "Dockerfile" to get the correct highlighting

File: docs/userguide/eng-image/dockerfile_best-practices.md [hunk 8]
```diff
@@ -134,6 +134,41 @@ image. We recommend the [Debian image](https://hub.docker.com/_/debian/)
...
+# Set one or more individual labels
+LABEL com.example.version="0.0.1-beta"
+LABEL vendor="ACME Incorporated"
+LABEL com.example.release-date="2015-02-12"
+LABEL com.example.version.is-production=""
+
+# Set multiple labels on one line
```
Comment (thaJeztah) [inline docs/userguide/eng-image/dockerfile_best-practices.md]: Would it be better to put these descriptions outside of the code blocks, instead of a comment, so that each example has it's own code block ? (Just thinking out loud)

File: docs/userguide/eng-image/dockerfile_best-practices.md [hunk 9]
```diff
@@ -134,6 +134,41 @@ image. We recommend the [Debian image](https://hub.docker.com/_/debian/)
...
+
+# Set multiple labels at once, using line-continuation characters to break long lines
+LABEL vendor=ACME\ Incorporated \
+      com.example.is-beta= \
+      com.example.is-production="" \
+      com.example.version="0.0.1-beta" \
+      com.example.release-date="2015-02-12"
+```
+
+See (Understanding Docker Object Labels)(../labels-custom-metadata.md) for
```
Comment (thaJeztah) [inline docs/userguide/eng-image/dockerfile_best-practices.md]: s/Understanding Docker Object Labels/Understanding Docker object labels/

(hm reading that a couple of times now; should we drop "Docker" in that, so "Understanding object labels" ? Again, thinking out loud)

File: docs/userguide/intro.md [hunk 10]
```diff
@@ -64,6 +65,25 @@ learning how to manage data, volumes and mounts inside our containers.
 
 Go to [Managing Data in Containers](../tutorials/dockervolumes.md).
 
+## Managing Metadata (Labels) for Docker Objects
+
+_Labels_ are a mechanism for applying metadata to Docker objects, including:
```
Comment (thaJeztah) [inline docs/userguide/intro.md]: Not sure we need the emphasis on "Labels" here, I think we can do without

File: docs/userguide/intro.md [hunk 11]
```diff
@@ -64,6 +65,25 @@ learning how to manage data, volumes and mounts inside our containers.
 
 Go to [Managing Data in Containers](../tutorials/dockervolumes.md).
 
+## Managing Metadata (Labels) for Docker Objects
```
Comment (thaJeztah) [inline docs/userguide/intro.md]: s/Managing Metadata (Labels) for Docker Objects/Managing metadata (labels) for Docker objects/

File: docs/userguide/intro.md [hunk 12]
```diff
@@ -64,6 +65,25 @@ learning how to manage data, volumes and mounts inside our containers.
...
+- Networks
+- Swarm nodes
+- Swarm services
+
+
+You can use labels to organize your images, record licensing information, annotate
+relationships between containers, volumes, and networks, or in any way that makes
+sense for your business or application.
+
+Go to [Managing Metadata \(Labels\) for Docker Objects](labels-custom-metadata.md).
```
Comment (thaJeztah) [inline docs/userguide/intro.md]: s/Managing Metadata (Labels) for Docker Objects/Managing metadata (labels) for Docker objects/

(Oh, surprised the parentheses needed escaping here)

Comment (thaJeztah) [inline docs/userguide/intro.md]: Actually, it looks like the link caption doesn't match the actual title of the document now

File: docs/userguide/labels-custom-metadata.md [hunk 13]
```diff
@@ -1,230 +1,119 @@
...
+keywords = ["Usage, user guide, labels, metadata, docker, documentation, examples, annotating"]
 [menu.main]
-parent = "engine_guide"
+parent = "engine_guide_intro"
 weight=90
 +++
 <![end-metadata]-->
 
-# Apply custom metadata
+# About Labels
```
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: s/Labels/labels/

File: docs/userguide/labels-custom-metadata.md [hunk 14]
```diff
@@ -1,230 +1,119 @@
...
 
-You can apply metadata to your images, containers, volumes, networks, nodes, services or daemons via
-labels. Labels serve a wide range of uses, such as adding notes or licensing
-information to an image, or to identify a host.
+_Labels_ are a mechanism for applying metadata to Docker objects, including:
```
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: same here w.r.t. emphasis on "labels"

File: docs/userguide/labels-custom-metadata.md [hunk 15]
```diff
@@ -1,230 +1,119 @@
...
 
->**Note:** Support for daemon-labels was added in Docker 1.4.1. Labels on
->containers and images were added in Docker 1.6.0
+You can use labels to organize your images, record licensing information, annotate
+relationships between containers, volumes, and networks, or in any way that makes
+sense for your business or application.
 
-## Label keys (namespaces)
+A label is a key-value pair, stored as a string. You can specify multiple labels
+to an object, but each key-value pair must be unique within an object. If the
```
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: is it "specify _to_"? (English is not my native language, so not sure)

File: docs/userguide/labels-custom-metadata.md [hunk 16]
```diff
@@ -1,230 +1,119 @@
...
+same key is given multiple values, the most-recently-written value overwrites
+all previous values.
 
-Docker puts no hard restrictions on the `key` used for a label. However, using
-simple keys can easily lead to conflicts. For example, you have chosen to
-categorize your images by CPU architecture using "architecture" labels in
-your Dockerfiles:
+>**Note:** To apply labels to Docker daemons, you need Docker Engine 1.4.1 or
+higher. To apply labels to images and containers, you need Docker Engine 1.6.0
+or higher.
```
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: Engine versions older than 1.6 are no longer supported by Docker Hub, so we can remove this note entirely

File: docs/userguide/labels-custom-metadata.md [hunk 17]
```diff
@@ -1,230 +1,119 @@
...
 
-    LABEL architecture="amd64"
+# Label keys and values
+## Key Format recommendations
```
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: s/Format/format/

File: docs/userguide/labels-custom-metadata.md [hunk 18]
```diff
@@ -1,230 +1,119 @@
...
 
-    LABEL architecture="ARMv7"
+A label _key_ is the left-hand side of the key-value pair. Keys are alphanumeric
+strings which may contain periods (.) and hyphens (-). Most Docker users use
+images created by other organizations, and the following guidelines help to
+prevent naming conflicts between Docker objects, especially if you plan to use
```
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: Bit unsure about "naming conflicts between Docker objects", but don't have a good suggestion for an alternative

File: docs/userguide/labels-custom-metadata.md [hunk 19]
```diff
@@ -1,230 +1,119 @@
...
 
-    LABEL architecture="Art Nouveau"
-
-To prevent naming conflicts, Docker recommends using namespaces to label keys
-using reverse domain notation. Use the following guidelines to name your keys:
-
-- All (third-party) tools should prefix their keys with the
-  reverse DNS notation of a domain controlled by the author. For
-  example, `com.example.some-label`.
+- Do not use a domain in your label key without the domain owner's permission.
```
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: 👍 nice

File: docs/userguide/labels-custom-metadata.md [hunk 20]
```diff
@@ -1,230 +1,119 @@
...
-
-    LABEL [<namespace>.]<key>=<value> ...
-
-The `LABEL` instruction adds a label to your image. A `LABEL` consists of a `<key>`
-and a `<value>`.
-Use an empty string for labels  that don't have a `<value>`,
-Use surrounding quotes or backslashes for labels that contain
-white space characters in the `<value>`:
-
-    LABEL vendor=ACME\ Incorporated
```
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: When I first saw all this being removed I thought "what?" but it's all duplicated in other parts of the docs! Nice cleanup 💯

File: docs/userguide/labels-custom-metadata.md [hunk 21]
```diff
@@ -1,230 +1,119 @@
...
-     com.example.environment=production
-     com.example.storage=ssd
+  reserved by Docker for internal use.
+
+- Label keys should begin and end with a lower-case letter and should only
+  contain lower-case alphanumeric characters, the period character (`.`), and
+  the hyphen character (`-`). Consecutive periods or hyphens are not allowed.
+
+- The period character (`.`) separates namespace "fields". Label keys without
+  namespaces are reserved for CLI use, allowing an end user to interactively
```
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: end-user ? (hyphen)

File: docs/userguide/labels-custom-metadata.md [hunk 22]
```diff
@@ -1,230 +1,119 @@
...
+## Value guidelines
+
+Label values can contain any data type that can be represented as a string,
+including (but not limited to) JSON, XML, CSV, or YAML. The only requirement is
+that the value be serialized to a string first, using a mechanism specific to
+the type of structure. For instance, to serialize JSON into a string, you might
+use the `JSON.stringify()` JavaScript method.
+
+Since Docker does not deserialize the value, you cannot treat a JSON or XML
+document as a nested stricture when querying or filtering by label value unless
```
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: typo: s/stricture/structure/

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks @mstanleyjones, looks great. Left comments / suggestions inline

File: docs/userguide/labels-custom-metadata.md (hunk 14 above)
Comment (mstanleyjones) [inline docs/userguide/labels-custom-metadata.md]: It's capitalized because it's the first word of the sentence. I italicized it to call it out as terminology at the first use. I think that's pretty standard.

**Comment by mstanleyjones** [comment]
Thanks for the comments. I think I've addressed them all.

File: docs/userguide/eng-image/dockerfile_best-practices.md (hunk 7 above)
Comment (mstanleyjones) [inline docs/userguide/eng-image/dockerfile_best-practices.md]: Yep, nohighlight is build into highlight.js already. I didn't know HLJS understood dockerfile, but I'll try it.

File: docs/userguide/eng-image/dockerfile_best-practices.md (hunk 8 above)
Comment (mstanleyjones) [inline docs/userguide/eng-image/dockerfile_best-practices.md]: I don't think so. It emphasizes a couple things -- comments in Dockerfiles and the fact that you can have multiple labels in a Dockerfile. If you feel strongly about it, I can change it.

File: docs/userguide/intro.md (hunk 12 above)
Comment (mstanleyjones) [inline docs/userguide/intro.md]: Good catch. I am not sure the parentheses needed escaping, but since () has a special meaning in Markdown links, I thought it might, and better safe than sorry. It's a moot point now.

File: docs/userguide/eng-image/dockerfile_best-practices.md [hunk 23]
```diff
@@ -134,6 +134,41 @@ image. We recommend the [Debian image](https://hub.docker.com/_/debian/)
 since it’s very tightly controlled and kept minimal (currently under 150 mb),
 while still being a full distribution.
 
+### LABEL
+
+(Understanding object labels)(../labels-custom-metadata.md)
```
Comment (thaJeztah) [inline docs/userguide/eng-image/dockerfile_best-practices.md]: s/(Understanding object labels)/[Understanding object labels]/ (square brackets)

File: docs/userguide/eng-image/dockerfile_best-practices.md (hunk 8 above)
Comment (thaJeztah) [inline docs/userguide/eng-image/dockerfile_best-practices.md]: I think it's ok here, no need to change. In some cases people overlook the comments *inside* code blocks, hence my suggestion

File: docs/userguide/eng-image/dockerfile_best-practices.md [hunk 24]
```diff
@@ -134,6 +134,41 @@ image. We recommend the [Debian image](https://hub.docker.com/_/debian/)
...
+
+# Set multiple labels at once, using line-continuation characters to break long lines
+LABEL vendor=ACME\ Incorporated \
+      com.example.is-beta= \
+      com.example.is-production="" \
+      com.example.version="0.0.1-beta" \
+      com.example.release-date="2015-02-12"
+```
+
+See (Understanding object labels)(../labels-custom-metadata.md) for
```
Comment (thaJeztah) [inline docs/userguide/eng-image/dockerfile_best-practices.md]: Same here (square brackets)

File: docs/userguide/eng-image/dockerfile_best-practices.md [hunk 25]
```diff
@@ -134,6 +134,41 @@ image. We recommend the [Debian image](https://hub.docker.com/_/debian/)
...
+guidelines about acceptable label keys and values. For information about
+querying labels, refer to ()[].
```
Comment (thaJeztah) [inline docs/userguide/eng-image/dockerfile_best-practices.md]: Link is missing here

File: docs/userguide/labels-custom-metadata.md (hunk 14 above)
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: I don't think we use that in other locations, but other than that, no big objections here

File: docs/userguide/labels-custom-metadata.md [hunk 26]
```diff
@@ -1,230 +1,115 @@
...
+A label is a key-value pair, stored as a string. You can specify multiple labels
+for an object, but each key-value pair must be unique within an object. If the
+same key is given multiple values, the most-recently-written value overwrites
+all previous values.
 
-Docker puts no hard restrictions on the `key` used for a label. However, using
-simple keys can easily lead to conflicts. For example, you have chosen to
-categorize your images by CPU architecture using "architecture" labels in
-your Dockerfiles:
+# Label keys and values
```
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: We could move this heading above the previous paragraph to prevent having two headings without text (think there was a nice writers term for that, but I forgot)

File: docs/userguide/labels-custom-metadata.md [hunk 27]
```diff
@@ -1,230 +1,115 @@
...
+## Key format recommendations
 
-    LABEL architecture="amd64"
+A label _key_ is the left-hand side of the key-value pair. Keys are alphanumeric
+strings which may contain periods (.) and hyphens (-). Most Docker users use
```
Comment (thaJeztah) [inline docs/userguide/labels-custom-metadata.md]: Put the period and hyphen in back tics here as well? (I see you did it below)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Found some Markdown syntax errors, and a couple of minor nits. Almost there! :smile:

**Comment by mstanleyjones** [comment]
Requested changes have been made, @thaJeztah

**Review by thaJeztah — APPROVED** [review]

**Comment by thaJeztah** [comment]
LGTM, thanks!
